src/drivers/intel/cavs/interrupt.c

```diff
@@ -73,7 +74,8 @@ static inline void irq_lvl2_handler(void *data, int level, uint32_t ilxsd,
 		list_for_item(clist, &cascade->child[i]) {
```

Comment:
We iterate over children list, but we do not do cache invalidation before.
You can have out of date data here, if some other core in the meantime registered/unregistered interrupt.

Reply:
On the one hand I'm not changing that, I'm basically just replacing one struct with a different one in this patch. But I think you're right. And it's even worse than that. It isn't just a caching problem, it's a real race: this handler isn't taking the spinlock and the child interrupt registration function isn't using spin_lock_irq(), it's only doing spin_lock(), so, while a process is registering an interrupt, if another handler is already registered for that interrupt, it can trigger, it will interrupt the registration code and can access the corrupted list. Thanks for spotting this!

Reply:
Actually, I think, we didn't need cache-flushing before my patches and also with the first of them, because there interrupt descriptors were per-core. It's only with the second patch in this series we move to a single descriptor for all cores, then we need cache-flushing.